### PR TITLE
Add fqtk as a demultiplexer

### PR DESCRIPTION
**Pull request by sam-white04**
Hello, 

I have added [fqtk](https://github.com/fulcrumgenomics/fqtk) as an optional demultiplexer. The tool fqtk requires an additional input to be provided as a path in the 5th column of `--input samplesheet.csv`.  

Thank you, 
Samantha White

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/demultiplex/tree/master/.github/CONTRIBUTING.md)- [x] If necessary, also make a PR on the nf-core/demultiplex _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated. (@sam-white04 Will update as soon as PR is posted)
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by Emiller88** [comment]
I think all of your commits got squashed 🙃

**Comment by sam-white04** [comment]
These checks passed locally. So let me do some digging to fix this.

File: README.md [hunk 1]
```diff
@@ -18,6 +18,7 @@
 1. Illumina (via `bcl2fastq` or `bclconvert`)
 2. Element Biosciences (via `bases2fastq`)
 3. Singular Genomics (via [`sgdemux`](https://github.com/Singular-Genomics/singular-demux))
+4. FASTQ files with user supplied read structures (via ['fqtk'](https://github.com/fulcrumgenomics/fqtk))
```
Comment (Emiller88) [inline README.md:21]:
```suggestion
4. FASTQ files with user supplied read structures (via [`fqtk`](https://github.com/fulcrumgenomics/fqtk))
```

File: subworkflows/local/fqtk_demultiplex/main.nf [hunk 2]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env nextflow
+
+//
+// Demultiplex FASTQ files with fqtk
+//
+
+include { FQTK }  from "../../../modules/nf-core/fqtk/main"
+include {CSV2TSV} from "../../../modules/local/csv2tsv"
```
Comment (Emiller88) [inline subworkflows/local/fqtk_demultiplex/main.nf:8]:
```suggestion
include { CSV2TSV } from "../../../modules/local/csv2tsv"
```

File: subworkflows/local/fqtk_demultiplex/main.nf [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    // "@<instrument>:<run number>:<flowcell ID>:<lane>:<tile>:<x-pos>:<y-pos>:UMI <read>:N:0:<index sequence>"
+    sequencer_serial = fields[0]
+    run_nubmer       = fields[1]
+    fcid             = fields[2]
+    lane             = fields[3]
+    index            = fields[-1] =~ /[GATC+-]/ ? fields[-1] : ""
+
+    rg.ID = [fcid,lane].join(".")
+    rg.PU = [fcid, lane, index].findAll().join(".")
+    rg.PL = "SINGULAR"
```
Comment (Emiller88) [inline subworkflows/local/fqtk_demultiplex/main.nf:95]: Should this be hard-coded?

File: workflows/demultiplex.nf [hunk 4]
```diff
@@ -73,35 +73,54 @@ include { MD5SUM                        } from '../modules/nf-core/md5sum/main'
...
-    ch_flowcells_tar = ch_flowcells.tar
-        .multiMap { meta, samplesheet, run ->
-            samplesheets: [ meta, samplesheet ]
-            run_dirs: [ meta, run ]
-        }
+        ch_flowcells_tar = ch_flowcells.tar
+            .multiMap { meta, samplesheet, run, manifest ->
+                samplesheets: [ meta, samplesheet, manifest ]
+                run_dirs: [ meta, run ]
+            }
```
Comment (Emiller88) [inline workflows/demultiplex.nf:105]: Could we put this inside of the switch case for fqtk instead? Maybe turn it into a subworkflow? Possibly turn them both into a subworkflow?

File: workflows/demultiplex.nf [hunk 5]
```diff
@@ -340,6 +380,48 @@ def parse_flowcell_csv(row) {
...
+            'per_flowcell_manifest': [
+                'content': 'path',
+                'pattern': '',
+            ]
+        ],
+        required: ['id','flowcell', 'samplesheet', 'per_flowcell_manifest'],
+    ]
+
+    return extract_csv(input_csv, input_schema)
+}
```
Comment (Emiller88) [inline workflows/demultiplex.nf:423]: These are starting to get excessive. @matthdsm what happens if we put them in `lib/`? Would they still work the same?

**Review by Emiller88 — CHANGES_REQUESTED** [review]
Few small comments and style changes, but looks pretty good!

File: workflows/demultiplex.nf (hunk 4 above)
Comment (sam-white04) [reply workflows/demultiplex.nf:105]: @Emiller88 I agree this gets a bit messy, Im happy to pull both into subworkflows where the input is `ch_input` and the output is `ch_flowcells` and `ch_flowcells_tar`. Thoughts?

File: workflows/demultiplex.nf (hunk 5 above)
Comment (sam-white04) [reply workflows/demultiplex.nf:423]: @Emiller88 Moving the prep for `ch_flowcells` into subworkflows would remove the need for these functions to be in `demultiplex.nf`, right?

**Review by sam-white04 — COMMENTED** [review]
@Emiller88 Thank you for your comments. If everyone is on the same page, Im happy to pull the generation of `ch_flowcells` and `ch_flowcells_tar` into subworkflows for fqtk vs the other demultiplexers.

File: subworkflows/local/fqtk_demultiplex/main.nf (hunk 3 above)
Comment (Emiller88) [reply subworkflows/local/fqtk_demultiplex/main.nf:95]: There might be a way to read in the read group. We can just throw a TODO on it though for now.

File: workflows/demultiplex.nf (hunk 5 above)
Comment (Emiller88) [reply workflows/demultiplex.nf:423]: Right, that's a great idea!

**Review by Emiller88 — APPROVED** [review]
I think this looks good to me. #102 For follow up to clean up some things that needed to be introduced here.

File: workflows/demultiplex.nf (hunk 5 above)
Comment (matthdsm) [reply workflows/demultiplex.nf:423]: I've no qualms putting all the functions in `/lib`. They're only in the main file for convenience!

**Review by matthdsm — APPROVED** [review]

**Pull request merged**